### PR TITLE
Feature/rand bucket

### LINKED ISSUE #3: Add random number for bucketing to TrustAtom
Feature: 

link tag when complete looks like:

`Ŧ→sushi[0x00]0.999999999[0x00]892412523[0x00]uhCEk…UFnFF`

This feature is to add the random bucket, eg:

`892412523`

AC:

- 9 digits total
- tests

Impl:

- use hkd rand bytes (integer 0-255) function, convert to integers (0-9) -- so we get two integers from each byte

### PR DESCRIPTION
**Pull request by dauphin3**
fn gen_bucket calls hdk random_bytes returning String which gets converted and put into chunks vec which gets call in create then passed through trust_atom_link_tag

fixes #3

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -7,7 +7,7 @@ let
     holochainVersionId = "v0_0_119";
     rustVersion = {
       track = "stable";
-      version = "1.57.0";
+      version = "1.57.0";  # outside of nix = 1.58
```
Comment (harlantwood) [inline default.nix:10]:
```suggestion
      version = "1.57.0";
```

File: zomes/hc_zome_trust_atom/Cargo.toml [hunk 2]
```diff
@@ -15,6 +15,7 @@ hdk = "=0.0.116"
 holo_hash = {version = "*", features = ["encoding"]}
 regex = "1"
 serde = "1"
+# rand = "0.8.5"
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/Cargo.toml:18]:
```suggestion
```

File: zomes/hc_zome_trust_atom/src/lib.rs [hunk 3]
```diff
@@ -10,6 +10,7 @@
 
 // #![warn(clippy::cargo)]
 
+// extern crate rand;
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/lib.rs:13]:
```suggestion
```

File: zomes/hc_zome_trust_atom/src/lib.rs [hunk 4]
```diff
@@ -42,31 +43,31 @@ pub struct QueryMineInput {
...
-#[hdk_extern]
-pub fn query_mine(input: QueryMineInput) -> ExternResult<Vec<TrustAtom>> {
-  trust_atom::query_mine(
-    input.target,
-    input.content_full,
-    input.content_starts_with,
-    input.min_rating,
-  )
-}
+// #[hdk_extern]
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/lib.rs:46]: why are these lines commented out?

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 5]
```diff
@@ -1,10 +1,11 @@
 #![allow(clippy::module_name_repetitions)]
 
-use regex::Regex;
-use std::collections::BTreeMap;
-
 use hdk::prelude::holo_hash::EntryHashB64;
 use hdk::prelude::*;
+use regex::Regex;
+use std::collections::BTreeMap;
+use std::convert::TryInto;
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:7]: I think we get this by default in 2021 edition

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 6]
```diff
@@ -48,7 +49,8 @@ pub fn create(
   target: EntryHash,
   content: &str,
   value: &str,
-  attributes: BTreeMap<String, String>,
+  bucket: Vec<u8>,
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:52]: I think bucket need not be passed in to create, can be created within it.

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 7]
```diff
@@ -113,7 +121,11 @@ fn validate_value(value_str: &str) -> ExternResult<()> {
 //   Ok(value)
 // }
 
-fn trust_atom_link_tag(link_direction: &LinkDirection, mut chunks: Vec<&str>) -> LinkTag {
+fn trust_atom_link_tag(
+  link_direction: &LinkDirection,
+  mut chunks: Vec<&str>,
+  bucket: Vec<u8>,
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:127]: let's pass bucket as the last one in the `chunks` vector

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 8]
```diff
@@ -125,102 +137,112 @@ fn trust_atom_link_tag(link_direction: &LinkDirection, mut chunks: Vec<&str>) ->
...
 
-pub fn query_mine(
-  target: Option<EntryHash>,
-  content_full: Option<String>,
-  content_starts_with: Option<String>,
-  min_value: Option<String>,
-) -> ExternResult<Vec<TrustAtom>> {
-  let agent_address: EntryHash = agent_info()?.agent_initial_pubkey.into();
+fn gen_bucket() -> ExternResult<Vec<u8>> {
+  let bucket = random_bytes(9)?.into_vec();
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:153]: this will be 9 rand bytes, but not the ascii bytes representing 0-9 (could be any).  Checking on the test, should catch this...

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 9]
```diff
@@ -420,6 +442,22 @@ mod tests {
...
+  //   // for item in &test_data {
+  //   //   let input: u8 = item.0;
+  //   //   let expected = item.1;
+  //   //   assert_eq!(normalize_value(value), expected);
+  //   // }
   // }
+
+  #[test]
+  fn test_gen_bucket() {
+    let test = "123456789";
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:459]: you need to call the function under test, something like:

```suggestion
    let test = gen_bucket();
```

File: zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs [hunk 10]
```diff
@@ -88,6 +88,7 @@ pub async fn test_create_trust_atom() {
   let relevant_link_string = String::from_utf8(relevant_link_bytes).unwrap();
   let expected_link_tag_string = format!("{}{}{}{}{}", "Ŧ", "→", "sushi", unicode_nul, "0.8");
   assert_eq!(relevant_link_string, expected_link_tag_string);
+  assert_eq!(relevant_link_string, expected_link_tag_string);
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs:91]:
```suggestion
```

File: zomes/hc_zome_trust_atom/src/trust_atom.rs (hunk 6 above)
Comment (dauphin3) [reply zomes/hc_zome_trust_atom/src/trust_atom.rs:52]: fixed

File: Cargo.lock [hunk 11]
```diff
@@ -1432,7 +1432,7 @@ dependencies = [
  "nonzero_ext",
  "parking_lot 0.11.2",
  "quanta",
- "rand 0.8.4",
```
Comment (harlantwood) [inline Cargo.lock:1435]: This file should be unchanged

Comment (harlantwood) [reply Cargo.lock:1435]: The whole cargo.lock file that is

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 12]
```diff
@@ -136,9 +159,9 @@ pub fn query_mine(
   Ok(result)
 }
 
-/// Required: exactly one of source or target
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:139]: Triple slashes are intentional — these create rust docs

File: zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs [hunk 13]
```diff
@@ -86,15 +86,24 @@ pub async fn test_create_trust_atom() {
...
 
   let chunks: Vec<&str> = relevant_link_string.split(unicode_nul).collect();
-  assert_eq!(chunks.len(), 2);
+  assert_eq!(chunks.len(), 3);
   assert_eq!(chunks[0], "Ŧ→sushi");
   assert_eq!(chunks[1], ".800000000");
 
+  let bucket = chunks[2];
+
+  println!("------> bucket: {:?}", bucket);
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs:97]: Delete line

File: zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs [hunk 14]
```diff
@@ -86,15 +86,24 @@ pub async fn test_create_trust_atom() {
...
+  assert_eq!(bucket.chars().count(), 9);
+  assert!(bucket.chars().all(|c| c.is_digit(10)));
+
+  let expected_link_tag_string = format!(
+    "{}{}{}{}{}{}{}",
+    "Ŧ", "→", "sushi", unicode_nul, ".800000000", unicode_nul, bucket
+  );
+  assert_eq!(relevant_link_string, expected_link_tag_string);
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs:105]: Excellent

File: zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs [hunk 15]
```diff
@@ -114,12 +123,14 @@ pub async fn test_create_trust_atom() {
...
-    format!("{}{}{}{}{}", "Ŧ", "↩", "sushi", unicode_nul, ".800000000");
+  let expected_link_tag_string = format!(
+    "{}{}{}{}{}{}{}",
+    "Ŧ", "↩", "sushi", unicode_nul, ".800000000", unicode_nul, bucket
+  );
   assert_eq!(relevant_link_string, expected_link_tag_string);
 
   let chunks: Vec<&str> = relevant_link_string.split(unicode_nul).collect();
-  assert_eq!(chunks.len(), 2);
+  assert_eq!(chunks.len(), 3);
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs:133]: After checking chunks 0 and 1 also check chunk 2 == bucket

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 16]
```diff
@@ -109,14 +115,31 @@ fn trust_atom_link_tag(link_direction: &LinkDirection, mut chunks: Vec<&str>) ->
...
+    let rand = random_bytes(1)?.into_vec();
+    let val = rand.get(0);
+    if let Some(n) = val {
+      bucket += (*n % 10).to_string().as_str();
+      total += 1;
+    } else {
+      return Err(WasmError::Guest("something went wrong".to_string()));
+    }
+  }
+  debug!("bucket: {:#?}", bucket);
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:139]: Delete debugs before merging

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 17]
```diff
@@ -109,14 +115,31 @@ fn trust_atom_link_tag(link_direction: &LinkDirection, mut chunks: Vec<&str>) ->
...
-  for chunk in &chunks {
+  for chunk in chunks {
     link_tag_bytes.extend_from_slice(&UNICODE_NUL_BYTES);
     link_tag_bytes.extend_from_slice(chunk.as_bytes());
   }
 
   LinkTag(link_tag_bytes)
 }
 
+fn gen_bucket() -> ExternResult<String> {
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:126]: I suggest create_bucket for the name.  

For algo, I suggest:

- get random_bytes(4)
- This gives you 4 ints each 0-255
- from these, derive a single int, in range 0 - 255^4
- take this % 10^9
- Now you have an int in range 0 - 999999999
- Format to a string, adding leading zeros if needed
- for this You can use format! With the right format args
- done 😎

Comment (harlantwood) [reply zomes/hc_zome_trust_atom/src/trust_atom.rs:126]: Also suggest we break this into 2 functions— one takes a vec of bytes, and does all the useful work and can be unit tested, the other calls hdk function. 

Then add a unit test which calls the first one 1000 times.

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 18]
```diff
@@ -109,14 +117,29 @@ fn trust_atom_link_tag(link_direction: &LinkDirection, mut chunks: Vec<&str>) ->
...
+  let bucket_bytes = random_bytes(9)?.into_vec();
+  Ok(create_bucket_string(&bucket_bytes))
+}
+
+fn create_bucket_string(bucket_bytes: &[u8]) -> String {
+  let mut bucket = String::new();
+  for chunk in bucket_bytes {
+    let val = chunk;
+    bucket += (val % 10).to_string().as_str();
+    // maybe: ( val.to_float / 25.5 ).to_int
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:138]: would be great to do something like this, pretty cheap and makes it feel more pro.  
would want to increase tests a bit more then; like all the boundary conditions:
0, 1, 24, 25, 26, 50, 51, 52, etc up to 255.

but  we could break it out into a separate task if you prefer.

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 19]
```diff
@@ -47,17 +47,25 @@ pub fn create(
...
   let agent_address: EntryHash = agent_info.agent_initial_pubkey.into();
 
   let normalized_value = normalize_value(value)?;
 
-  let forward_link_tag =
-    trust_atom_link_tag(&LinkDirection::Forward, vec![content, &normalized_value]);
-  let reverse_link_tag =
-    trust_atom_link_tag(&LinkDirection::Reverse, vec![content, &normalized_value]);
+  // let bucket_bytes = generate_random_bucket()?;
+  // let bucket = create_bucket_string(bucket_bytes);
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:58]:
```suggestion
```

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 20]
```diff
@@ -136,9 +159,9 @@ pub fn query_mine(
   Ok(result)
 }
 
-/// Required: exactly one of source or target
-/// All other arguments are optional
-/// Arguments act as additive filters (AND)
+// Required: exactly one of source or target
+// All other arguments are optional
+// Arguments act as additive filters (AND)
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:164]:
```suggestion
/// Required: exactly one of source or target
/// All other arguments are optional
/// Arguments act as additive filters (AND)
```

File: zomes/hc_zome_trust_atom/src/trust_atom.rs (hunk 18 above)
Comment (harlantwood) [reply zomes/hc_zome_trust_atom/src/trust_atom.rs:136]: Moved to #17
